_posts/2020-10-05-how-to-write-a-new-foreman-ansible-module-in-20-lines-of-code.md

```diff
@@ -0,0 +1,161 @@
+    locations: "{{ http_proxy_locations | default(omit) }}"
+    organizations: "{{ http_proxy_organizations | default(omit) }}"
+    state: "{{ http_proxy_state }}"
+  register: result
+- assert:
+    fail_msg: "Ensuring http_proxy is {{ http_proxy_state }} failed! (expected_change: {{ expected_change | default('unknown') }})"
+    that:
+      - result.changed == expected_change
+  when: expected_change is defined
+...
+```
+
+In this task file, there are two tasks. The first one calls the `http_proxy` module, passing in the connection information and the module-specific parameters, most of which have `default(omit)` specified, which means these parameters won't be passed to the module if the variable isn't set -- super useful when you need to test the module with different sets of parameters. And the second one that ensures that the module reported the `changed` attribute correctly.
+
+Now the main playbook (`http_proxy.yml`) is a bit longer, so we won't paste it verbatim here, but discuss the individual sections.
+
+First, it has a play with `hosts: localhost` to prepare the test environment: setup organizations, locations and whatever else might be needed.
+
+This is followed by a long play with `hosts: tests` that contains the actual test steps.
+Those steps follow a pattern: do something and expect a change to happen (`expected_change: true`), do the same thing again and expect no change to happen (`expected_change: false`). This ensures idempotency of the module.
+
+The "things to do" again follow a pattern: create a proxy with minimal information, update a proxy, delete a proxy. This ensures the common workflows of a user are covered.
+
+And last, it has is another play with `hosts: localhost` to tear down anything that has been prepared for this specific test to run.
+
+The differentiation between `hosts: localhost` and `hosts: tests` allows us to execute the `localhost` steps only when we're recording a new set of API interactions and the `tests` steps always, as there is no need for the setup and teardown when running the tests against the recorded data.
+
+The last missing piece between us and working tests is a symlink. To keep the recorded API data clean of the big apidoc JSON and also allow easier updates when the apidoc changes, we keep copies of the JSON in `tests/fixtures/apidoc`. However, we keep only `foreman.json`, `katello.json` and `luna.json` as real files while all the others are symlinks. As the HTTP Proxy is a Foreman feature, we will create a symlink from `http_proxy.json` to `foreman.json`. Would it have been a Katello feature, the link would have been to `katello.json`. And for any other plugin, it's `luna.json` as that's essentially "Foreman with all the nice plugins".
+
+Having all pieces in place (and a running Foreman instance somewhere), we can run `make record_http_proxy` which it will execute our playbook and record the API interactions in `tests/test_playbooks/fixtures`.
```

Comment:
Ideas:
"Don't forget to check those fixtures in when submitting a PR."
"You can rerecord at any time."
"Use `make livetest_http_proxy` to test your playbook locally."
"The recorded fixtures are verified with `make test_http_proxy`."